eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,399 @@
+		// Getting historical logs may take significant amount of time.
+		// The current client may fail in the middle of the process,
+		// and also some clients may become available again, therefore we need to try other clients on error.
+		// Since singleLogsCh and singleErrCh are initialized by the active client,
+		// we need to re-initialize them if we try other clients.
+		processLogsFunc := func(client SingleClientProvider) (any, error) {
+			if needInitChans.Load() {
+				if _, err := startFetchingFunc(client); err != nil {
+					return nil, err
+				}
+			} else {
+				needInitChans.Store(true)
+			}
+
+			var lastBlock uint64
+			for {
+				select {
+				case <-ctx.Done():
+					return nil, ctx.Err()
+				case <-ec.closed:
+					return nil, fmt.Errorf("client closed")
+				case log, ok := <-singleLogsCh:
+					if !ok {
+						// Underlying channel is closed -> no more logs.
+						return nil, nil
+					}
+					logsCh <- log
+					lastBlock = max(lastBlock, log.BlockNumber)
+				case err, ok := <-singleErrCh:
+					if !ok {
+						// If the error channel closed, treat that as no more logs or success.
+						return nil, nil
+					}
+					fromBlock = max(fromBlock, lastBlock+1)
+					return nil, err // Try another client
+				}
+			}
+		}
+
+		if _, err := ec.call(ec.setMethod(ctx, "FetchHistoricalLogs [process logs]"), processLogsFunc); err != nil {
+			errCh <- err
+		}
+	}()
+
+	return logsCh, errCh, nil
+}
+
+// StreamLogs subscribes to events emitted by the contract.
+func (ec *MultiClient) StreamLogs(ctx context.Context, fromBlock uint64) <-chan BlockLogs {
+	logs := make(chan BlockLogs)
+
+	go func() {
+		defer close(logs)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ec.closed:
+				return
+			default:
+				f := func(client SingleClientProvider) (any, error) {
+					lastBlock, err := client.streamLogsToChan(ctx, logs, fromBlock)
+					if errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled) {
+						return lastBlock, err
+					}
+					if err != nil {
+						// fromBlock's value in the outer scope is updated here, so this function needs to be a closure
+						fromBlock = max(fromBlock, lastBlock+1)
+						return nil, err
+					}
+
+					return nil, nil
+				}
+
+				_, err := ec.call(ec.setMethod(ctx, "StreamLogs"), f)
+				if err != nil && !errors.Is(err, ErrClosed) && !errors.Is(err, context.Canceled) {
+					// NOTE: There are unit tests that trigger Fatal and override its behavior.
+					// Therefore, the code must call `return` afterward.
+					ec.logger.Fatal("failed to stream registry events", zap.Error(err))
+				}
+				return
+			}
+		}
+	}()
+
+	return logs
+}
+
+// Healthy returns if execution client is currently healthy: responds to requests and not in the syncing state.
+func (ec *MultiClient) Healthy(ctx context.Context) error {
+	f := func(client SingleClientProvider) (any, error) {
+		return nil, client.Healthy(ctx)
+	}
+
+	if _, err := ec.call(ec.setMethod(ctx, "Healthy"), f); err != nil {
```

Comment:
IIRC it's a good standard in Go to have this as a func (rather than method) called `withMethod` and the key as an empty struct type

example: https://mainfunction.org/articles/golang-context-with-value-best-practices/

(we don't need the FromContext method)

Reply:
if withMethod conflicts with option funcs in this package (With<Option>) then we could maybe just call it `contextWithMethod` to be explicit

Reply:
Yes, I agree, it doesn't have to be a method. Why don't we need the `FromContext` method? It makes extracting value easier